eclipse/buildship
# Use "test" classpath attribute to separate main and test classpath

This PR makes use of Eclipse's new test sources feature. Since Gradle 5.6, the `EclipseProject` model [contains](https://github.com/gradle/gradle/pull/9484) the `test` classpath attribute on all source folders and dependencies that belong to a test configuration. 

Gradle 5.6 still [provides](https://github.com/gradle/gradle/pull/9789) the custom scope information, which is used by previous Buildship releases. Buildship falls back to that information if an older Gradle version is being used or if the Host Eclipse version doesn't support test sources (< [4.8](https://www.eclipse.org/eclipse/news/4.8/jdt.php#jdt-test-sources)).

### Testing the feature

1. Install the Eclipse installer: go to https://www.eclipse.org/downloads/ and click the `Download 64` button and proceed with the installer installation.
2. Provision the Buildship IDE
  - Open the Eclipse installer
  - Select `Advanced mode...` in the menu in the top-right corner

<img width="858" alt="Screenshot 2019-07-31 at 14 27 50" src="https://user-images.githubusercontent.com/419883/62214328-20a26480-b3a5-11e9-9777-6839450e9108.png">

  - On the product page select Eclipse IDE for Eclipse Committers and change the product version to the latest release

<img width="820" alt="Screenshot 2019-07-31 at 14 30 09" src="https://user-images.githubusercontent.com/419883/62214378-3dd73300-b3a5-11e9-9883-ae918b2bd17a.png">

  - On the Projects page select `Eclipse Projects` > `Buildship`

<img width="821" alt="Screenshot 2019-07-31 at 14 30 49" src="https://user-images.githubusercontent.com/419883/62214391-462f6e00-b3a5-11e9-89ef-5092bbc82258.png">

  - On the variables page set the git branch to `donat/test-sources`. Also, set the installation / workspace / git clone location to your liking. 

<img width="822" alt="Screenshot 2019-07-31 at 14 46 58" src="https://user-images.githubusercontent.com/419883/62214404-4cbde580-b3a5-11e9-84b1-3c0f9c4c7b66.png">

  - Click finish and wait for the provisioning to finish. Once done, you should see a fully configured IDE with no compilation errors.

3. Launch development version of Buildship to verify changes
 -  In Eclipse, Open the Run Configuration dialog and run the `Launch Buildship` configuration

<img width="1025" alt="Screenshot 2019-07-31 at 15 01 08" src="https://user-images.githubusercontent.com/419883/62214426-58a9a780-b3a5-11e9-98ae-cc38240ab8c0.png">

 - Create a new Gradle project, try to use a test class in the main sources.
 - You can verify the classpath content by enabling tracing in the `Launch Buildship` configuration.

<img width="1069" alt="Screenshot 2019-07-31 at 15 04 18" src="https://user-images.githubusercontent.com/419883/62214455-63643c80-b3a5-11e9-9db2-969aefdd59b0.png">

In the development IDE if you right-click on a Gradle project and select `Gradle > Refresh Gradle project` then you should see the classpath attributes in the host IDE.
